ocaml/testsuite/tests/typing-modes/portable-contend.ml

```diff
@@ -0,0 +1,249 @@
+Error: Found a nonportable value where a portable value was expected
+|}]
+
+(* Closing over reading mutable field gives nonportable *)
+let foo () =
+    let r = {a = "foo"; b = "bar"} in
+    let bar () = let _ = r.a in () in
+    let _ @ portable = bar in
+    ()
+[%%expect{|
+Line 4, characters 23-26:
+4 |     let _ @ portable = bar in
+                           ^^^
+Error: Found a nonportable value where a portable value was expected
+|}]
+
+(* Closing over reading immutable field is OK *)
+let foo () =
+    let r @ portable = {a = "foo"; b = "bar"} in
+    let bar () = let _ = r.b in () in
+    let _ @ portable = bar in
+    ()
+(* CR zqian: currently mutable(legacy) means all records constructed are nonportable,
+   and the above bar is closing over an nonportable record. Once we allow mutable()
+   syntax, we can test this. *)
+[%%expect{|
+Line 2, characters 23-45:
+2 |     let r @ portable = {a = "foo"; b = "bar"} in
+                           ^^^^^^^^^^^^^^^^^^^^^^
+Error: Found a nonportable value where a portable value was expected
+|}]
+
+
+(* TESTING arrays *)
+(* reading/writing to array requires uncontended *)
+let foo (r @ contended) = Array.set r 42 "hello"
+[%%expect{|
+Line 1, characters 36-37:
+1 | let foo (r @ contended) = Array.set r 42 "hello"
+                                        ^
+Error: Found a contended value where a uncontended value was expected
+|}]
+let foo (r @ contended) = Array.get r 42
+[%%expect{|
+Line 1, characters 36-37:
+1 | let foo (r @ contended) = Array.get r 42
+                                        ^
+Error: Found a contended value where a uncontended value was expected
+|}]
+let foo (r @ contended) =
+    match r with
+    | [| x; y |] -> ()
+[%%expect{|
+Line 3, characters 6-16:
+3 |     | [| x; y |] -> ()
+          ^^^^^^^^^^
+Error: Found a contended value where a uncontended value was expected
+|}]
+
+
+(* Closing over write gives nonportable *)
+let foo () =
+    let r = [| "hello"; "world" |] in
+    let bar () = Array.set r 0 "foo" in
+    let _ @ portable = bar in
+    ()
+[%%expect{|
+Line 4, characters 23-26:
+4 |     let _ @ portable = bar in
+                           ^^^
+Error: Found a nonportable value where a portable value was expected
+|}]
+
+(* Closing over read gives nonportable *)
+let foo () =
+    let r = [| "hello"; "world" |] in
+    let bar () = Array.get r 0 in
+    let _ @ portable = bar in
+    ()
+[%%expect{|
+Line 4, characters 23-26:
+4 |     let _ @ portable = bar in
+                           ^^^
+Error: Found a nonportable value where a portable value was expected
+|}]
+
+
+(* Closing over measuring length doesn't force nonportable *)
+let foo () =
+    let r @ portable = [| "hello"; "world" |] in
+    let bar () = Array.length r in
+    let _ @ portable = bar in
+    ()
+(* CR zqian: The follows fails, because currently all records are constructed as
+    nonportable. Once we support mutable() syntax, we can test this. *)
```

Comment:
This test wants to show that closing over `contended` access to an `uncontended` value does not force `nonportable`. But we are not able to show that, because `r` is `nonportable` which  `bar` also closes over and are forced `nonportable`.

Array carrys an implicit `mutable(nonportable)`, and as a result, they can only be constructed as `nonportable`.